File: src/four_sum/FourSum.java
package four_sum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by gavingeng on 2018-06-11 11:35.
 */
public class FourSum {
	
	public static List<List<Integer>> fourSum(int[] nums, int target) {
		Arrays.sort(nums);
		List<List<Integer>> retList = new ArrayList<>();
		
		for (int i = 0; i < nums.length - 1; i++) {
			if (i > 0 && nums[i] == nums[i - 1])
				continue;
			for (int j = i + 1; j < nums.length - 1; j++) {
				if (j > i + 1 && nums[j] == nums[j - 1])
					continue;
				int k = j + 1;
				int m = nums.length - 1;
				
				while (k < m) {
					int sum = nums[i] + nums[j] + nums[k] + nums[m];
					if (sum < target) {
						k++;
						while (nums[k] == nums[k - 1] && k < m)
							k++;
					} else if (sum > target) {
						m--;
						while (nums[m] == nums[m + 1] && k < m)
							m--;
					} else {
						retList.add(Arrays.asList(nums[i], nums[j], nums[k], nums[m]));
						k++;
						m--;
						while (nums[k] == nums[k - 1] && k < m)
							k++;
						while (nums[m] == nums[m + 1] && k < m)
							m--;
					}
					
				}
				
			}
			
		}
		return retList;
	}
	
	public static void main(String[] args) {
//		int[] nums = {1, 0, -1, 0, -2, 2};
//		int[] nums = {-3, -1, 0, 2, 4, 5};
//		int[] nums = {0, 0, 0, 0};
//		int[] nums = {-3, -2, -1, 0, 0, 1, 2, 3};
//		int target = 0;
		
		int[] nums = {-1, 2, 2, -5, 0, -1, 4};
		int target = 3;
		List<List<Integer>> retList = fourSum(nums, target);
		for (List<Integer> inner : retList) {
			StringBuffer buffer = new StringBuffer();
			for (Integer i : inner) {
				buffer.append(i + "\t");
			}
			System.out.println(buffer.toString());
		}
	}
}

File: src/four_sum/four_sum.py
#!/usr/bin/env python
# coding:utf-8
# filename:four_sum.py

'''
author:	gavingeng
date:	2018-06-11 11:34:28 
'''


class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        retList = list()

        for i in range(len(nums) - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, len(nums) - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                k = j + 1
                m = len(nums) - 1

                while k < m:
                    sum = nums[i] + nums[j] + nums[k] + nums[m]
                    if sum < target:
                        k += 1
                        while k < m and nums[k] == nums[k - 1]:
                            k += 1
                    elif sum > target:
                        m -= 1
                        while k < m and nums[m] == nums[m + 1]:
                            m -= 1
                    elif sum == target:
                        inner = [nums[i], nums[j], nums[k], nums[m]]
                        retList.append(inner)
                        k += 1
                        m -= 1
                        while k < m and nums[k] == nums[k - 1]:
                            k += 1
                        while k < m and nums[m] == nums[m + 1]:
                            m -= 1

        return retList


def test():
    # nums = [1, 0, -1, 0, -2, 2]
    # nums=[-3, -1, 0, 2, 4, 5]
    # nums=[0, 0, 0, 0]
    nums = [-3, -2, -1, 0, 0, 1, 2, 3]
    target = 0
    s = Solution()
    retList = s.fourSum(nums, target)
    for inner in retList:
        print inner


def test2():
    nums = [-1, 2, 2, -5, 0, -1, 4]
    target = 3
    s = Solution()
    retList = s.fourSum(nums, target)
    for inner in retList:
        print inner


def main():
    test()
    print "*" * 20
    test2()


if __name__ == '__main__':
    main()
